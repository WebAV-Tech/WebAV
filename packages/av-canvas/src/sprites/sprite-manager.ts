import { VisibleSprite, EventTool } from '@webav/av-cliper';

export enum ESpriteManagerEvt {
  ActiveSpriteChange = 'activeSpriteChange',
  AddSprite = 'addSprite',
}

export class SpriteManager {
  #sprites: VisibleSprite[] = [];

  #activeSprite: VisibleSprite | null = null;

  #evtTool = new EventTool<{
    [ESpriteManagerEvt.AddSprite]: (s: VisibleSprite) => void;
    [ESpriteManagerEvt.ActiveSpriteChange]: (s: VisibleSprite | null) => void;
  }>();

  on = this.#evtTool.on;

  get activeSprite(): VisibleSprite | null {
    return this.#activeSprite;
  }
  set activeSprite(s: VisibleSprite | null) {
    if (s === this.#activeSprite) return;
    this.#activeSprite = s;
    this.#evtTool.emit(ESpriteManagerEvt.ActiveSpriteChange, s);
  }

  /**
   * 添加 {@link VisibleSprite}
   * @param args {@link VisibleSprite}
   * @returns
   * @example
   const sprite = new VisibleSprite(
      new ImgClip({
        type: 'image/gif',
        stream: (await fetch('https://xx.gif')).body!,
      }),
    );
   */
  /**
   * Adds a {@link VisibleSprite}.
   * @param vs The VisibleSprite to add.
   * @returns A promise that resolves when the sprite is added.
   * @example
   const sprite = new VisibleSprite(
      new ImgClip({
        type: 'image/gif',
        stream: (await fetch('https://xx.gif')).body!,
      }),
    );
   await avCvs.addSprite(sprite);
   */
  async addSprite(vs: VisibleSprite): Promise<void> {
    await vs.ready;
    this.#sprites.push(vs);
    this.#sprites = this.#sprites.sort((a, b) => a.zIndex - b.zIndex);
    vs.on('propsChange', (props) => {
      if (props.zIndex == null) return;
      this.#sprites = this.#sprites.sort((a, b) => a.zIndex - b.zIndex);
    });

    this.#evtTool.emit(ESpriteManagerEvt.AddSprite, vs);
  }

  /**
   * 删除 {@link VisibleSprite}
   * @param args
   * @returns
   * @example
   const sprite = new VisibleSprite();
   avCvs.removeSprite(sprite);
   */
  /**
   * Removes a {@link VisibleSprite}.
   * @param spr - The sprite to be removed.
   * @example
   const sprite = new VisibleSprite();
   avCvs.removeSprite(sprite);
   */
  removeSprite(spr: VisibleSprite): void {
    if (this.#activeSprite === spr) this.activeSprite = null;
    this.#sprites = this.#sprites.filter((s) => s !== spr);
    spr.destroy();
  }

  getSprites(filter: { time: boolean } = { time: true }): VisibleSprite[] {
    return this.#sprites.filter(
      (s) =>
        s.visible &&
        (filter.time
          ? this.#renderTime >= s.time.offset &&
            this.#renderTime <= s.time.offset + s.time.duration
          : true),
    );
  }

  #renderTime = 0;
  updateRenderTime(time: number) {
    this.#renderTime = time;

    // 避免素材不可见，但渲染了素材边框控制点
    const as = this.activeSprite;
    if (
      as != null &&
      (time < as.time.offset || time > as.time.offset + as.time.duration)
    ) {
      this.activeSprite = null;
    }
  }

  destroy(): void {
    this.#evtTool.destroy();
    this.#sprites.forEach((s) => s.destroy());
    this.#sprites = [];
  }
}
