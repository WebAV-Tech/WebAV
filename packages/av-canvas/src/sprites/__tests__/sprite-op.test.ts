import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { SpriteManager } from '../sprite-manager';
import { createEl } from '../../utils';
import { draggabelSprite } from '../sprite-op';
import { MockVisibleSprite, crtMSEvt4Offset } from '../../__tests__/test-utils';

const cvsRatio = { w: 1, h: 1 };
let sprMng = new SpriteManager();

let cvsEl: HTMLCanvasElement;
beforeEach(() => {
  sprMng = new SpriteManager();
  cvsEl = createEl('canvas') as HTMLCanvasElement;
  cvsEl.style.cssText = 'width: 1280px; height: 720px';
  cvsEl.width = 1280;
  cvsEl.height = 720;
  document.body.appendChild(cvsEl);
});
afterEach(() => {
  cvsEl.remove();
});

describe('draggabelSprite', () => {
  test('canvas on mousedown', () => {
    const spyAEL = vi.spyOn(cvsEl, 'addEventListener');
    const spyREL = vi.spyOn(cvsEl, 'removeEventListener');

    const clear = draggabelSprite(cvsEl, sprMng, document.body);
    expect(spyAEL).toBeCalledWith('mousedown', expect.any(Function));
    expect(clear).toBeInstanceOf(Function);

    clear();
    expect(spyREL).toBeCalledWith('mousedown', expect.any(Function));
  });

  test('window on mouse event', async () => {
    const spyAEL = vi.spyOn(window, 'addEventListener');
    const spyREL = vi.spyOn(window, 'removeEventListener');
    const vs = new MockVisibleSprite();
    vi.spyOn(vs.rect, 'checkHit').mockReturnValue(true);
    await sprMng.addSprite(vs);
    sprMng.activeSprite = vs;
    const clear = draggabelSprite(cvsEl, sprMng, document.body);
    cvsEl.dispatchEvent(new MouseEvent('mousedown'));

    expect(spyAEL).toBeCalledTimes(2);
    expect(spyAEL).toHaveBeenNthCalledWith(
      1,
      'mousemove',
      expect.any(Function),
    );
    expect(spyAEL).toHaveBeenNthCalledWith(2, 'mouseup', expect.any(Function));

    clear();
    expect(spyREL).toHaveBeenNthCalledWith(
      1,
      'mousemove',
      expect.any(Function),
    );
    expect(spyREL).toHaveBeenNthCalledWith(2, 'mouseup', expect.any(Function));
  });

  test('move sprite', async () => {
    const vs = new MockVisibleSprite();
    vs.rect.x = 100;
    vs.rect.y = 100;
    vs.rect.w = 100;
    vs.rect.h = 100;

    await sprMng.addSprite(vs);
    sprMng.activeSprite = vs;

    const clear = draggabelSprite(cvsEl, sprMng, document.body);
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 110, 110));

    window.dispatchEvent(
      new MouseEvent('mousemove', {
        clientX: 100,
        clientY: 100,
      }),
    );
    expect(vs.rect.x).toBe(200);
    expect(vs.rect.y).toBe(200);

    // 鼠标移动超出边界
    window.dispatchEvent(
      new MouseEvent('mousemove', {
        clientX: 10000,
        clientY: 10000,
      }),
    );
    // 5% 安全边界
    expect(vs.rect.x).toBe(cvsEl.width - cvsEl.width * 0.05);
    expect(vs.rect.y).toBe(cvsEl.height - cvsEl.height * 0.05);

    clear();
  });
});

describe('scale sprite', () => {
  test('drag right ctrl', async () => {
    const vs = new MockVisibleSprite();
    await sprMng.addSprite(vs);
    sprMng.activeSprite = vs;
    vs.rect.w = 100;
    vs.rect.h = 100;

    // 激活 sprite
    const clear = draggabelSprite(cvsEl, sprMng, document.body);
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 0, 0));
    expect(sprMng.activeSprite).toBe(vs);

    window.dispatchEvent(new MouseEvent('mouseup'));
    // 命中 right ctrl
    cvsEl.dispatchEvent(
      crtMSEvt4Offset('mousedown', 100 * cvsRatio.w, 50 * cvsRatio.h),
    );
    window.dispatchEvent(
      new MouseEvent('mousemove', {
        clientX: 100,
        clientY: 100,
      }),
    );
    // 拖拽 right ctrl 缩放 rect 的宽度
    expect(vs.rect.w).toBe(100 + 100 / cvsRatio.w);

    clear();
  });

  test('drag rb(bottom right) ctrl', async () => {
    const vs = new MockVisibleSprite();
    await sprMng.addSprite(vs);
    sprMng.activeSprite = vs;
    vs.rect.w = 100;
    vs.rect.h = 100;

    // 激活 sprite
    const clear = draggabelSprite(cvsEl, sprMng, document.body);
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 0, 0));
    expect(sprMng.activeSprite).toBe(vs);

    window.dispatchEvent(new MouseEvent('mouseup'));
    // 命中 right ctrl
    cvsEl.dispatchEvent(
      crtMSEvt4Offset('mousedown', 100 * cvsRatio.w, 100 * cvsRatio.h),
    );
    window.dispatchEvent(
      new MouseEvent('mousemove', {
        clientX: 100,
        clientY: 100,
      }),
    );
    expect(vs.rect).toMatchSnapshot();

    clear();
  });
});

describe('rotate sprite', () => {
  test('rotate sprite', async () => {
    const vs = new MockVisibleSprite();
    await sprMng.addSprite(vs);
    sprMng.activeSprite = vs;
    vs.rect.w = 100;
    vs.rect.h = 100;

    // 激活 sprite
    const clear = draggabelSprite(cvsEl, sprMng, document.body);
    cvsEl.dispatchEvent(crtMSEvt4Offset('mousedown', 0, 0));
    expect(sprMng.activeSprite).toBe(vs);

    window.dispatchEvent(new MouseEvent('mouseup'));
    // 命中 rotate ctrl
    const {
      center,
      ctrls: { rotate },
    } = vs.rect;
    cvsEl.dispatchEvent(
      crtMSEvt4Offset(
        'mousedown',
        (rotate.x + center.x) * cvsRatio.w,
        (rotate.y + center.y) * cvsRatio.h,
      ),
    );
    window.dispatchEvent(
      new MouseEvent('mousemove', {
        clientX: 100,
        clientY: 100,
      }),
    );
    expect(vs.rect.angle).toMatchSnapshot();

    window.dispatchEvent(
      new MouseEvent('mousemove', {
        clientX: 100,
        clientY: 200,
      }),
    );
    expect(vs.rect.angle).toMatchSnapshot();

    clear();
  });
});
