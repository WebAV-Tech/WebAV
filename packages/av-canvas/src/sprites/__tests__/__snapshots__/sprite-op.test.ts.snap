// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rotate sprite > rotate sprite 1`] = `2.356194490192345`;

exports[`rotate sprite > rotate sprite 2`] = `2.819842099193151`;

exports[`scale sprite > drag rb(bottom right) ctrl 1`] = `
O {
  "fixedAspectRatio": false,
  "master": null,
  "on": [Function],
}
`;
